Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -345,28 +437,120 @@ void CGContextSynchronize(CGContextRef context) {
+    // https://developer.apple.com/library/content/documentation/GraphicsImaging/Conceptual/drawingwithquartz2d/dq_trans_layers/dq_trans_layers.html#//apple_ref/doc/uid/TP30001066-CH210-SW1
+    // Doing so would cause:
+    // * Double shadows
+    // * Double transparency
+    // * Double blending
+    // * Double clipping
+    auto& newDrawingState = newLayer.CurrentGState();
+    newDrawingState.alpha = 1.0;
+    newDrawingState.shadowColor = { 0.f, 0.f, 0.f, 0.f };
+    // newDrawingState.blendMode = kCGBlendModeNormal; // TODO GH#1389
+    newDrawingState.clippingGeometry = nullptr;
+    if (rect) {
+        CGRect transformedClippingRegion = CGContextConvertRectToDeviceSpace(this, *rect);
+
+        if (gStateForComposingLayer.HasShadow()) {
+            // The reference platform grows the layer's clipping region by the shadow's offset and its approximate width.
+            CGFloat xdelta = fabsf(gStateForComposingLayer.shadowOffset.width);
+            xdelta += gStateForComposingLayer.shadowBlur + 1;
+
+            CGFloat ydelta = fabsf(gStateForComposingLayer.shadowOffset.height);
+            ydelta += gStateForComposingLayer.shadowBlur + 1;
+
+            transformedClippingRegion = CGRectInset(transformedClippingRegion, -xdelta, -ydelta);
+        }
+
+        ComPtr<ID2D1RectangleGeometry> rectGeometry;
+        RETURN_IF_FAILED(Factory()->CreateRectangleGeometry(__CGRectToD2D_F(transformedClippingRegion), &rectGeometry));
+        RETURN_IF_FAILED(newDrawingState.IntersectClippingGeometry(rectGeometry.Get(), kCGPathEOFill));
+    }
+
+    // A GState's global alpha is used when per-primitive drawing is not an option (such as when
+    // composing down a full transparency layer.)
+    gStateForComposingLayer.globalAlpha = gStateForComposingLayer.alpha;
+    gStateForComposingLayer.alpha = 1.0;
+    return S_OK;
+}
+
+HRESULT __CGContext::PopLayer() {
+    if (_layerStack.size() <= 1) {
+        TraceError(TAG, L"Invalid attempt to pop base layer.");
+        return E_BOUNDS;
+    }
+
+    auto outgoingLayer = std::move(_layerStack.top());
+
+    ComPtr<ID2D1Image> outgoingImageTarget;
+    RETURN_IF_FAILED(outgoingLayer.GetTarget(&outgoingImageTarget));
+
+    ComPtr<ID2D1CommandList> outgoingCommandList;
+    if (FAILED(outgoingImageTarget.As(&outgoingCommandList))) {
+        TraceError(TAG, L"Popped a layer that was NOT a command list. PANIC!");
+        return E_UNEXPECTED;
+    }
+
+    RETURN_IF_FAILED(outgoingCommandList->Close());
+
+    // Destroy the top layer and all its state.
+    _layerStack.pop();
+
+    auto& incomingLayer = _layerStack.top();
+    ComPtr<ID2D1Image> incomingImageTarget;
+    RETURN_IF_FAILED(incomingLayer.GetTarget(&incomingImageTarget));
+
+    deviceContext->SetTarget(incomingImageTarget.Get());
+
+    RETURN_IF_FAILED(DrawImage(outgoingCommandList.Get()));
+
+    // Restore the previous D2D state block & drawing parameters from the incoming layer.
+    RETURN_IF_FAILED(PopGState());
```

Comment:
And what if you add something after PopGState? It introduces unnecessary churn to combine the return with the erstwhile final statement.

---
In reply to: [90589430](https://github.com/Microsoft/WinObjC/pull/1496#discussion_r90589430) [](ancestors = 90589430)